delfin/drivers/dell_emc/vmax/rest.py

```diff
@@ -0,0 +1,548 @@
+# Copyright 2020 The SODA Authors.
+# Copyright (c) 2020 Dell Inc. or its subsidiaries.
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import json
+import sys
+
+from oslo_log import log as logging
+import requests
+import urllib3
+import requests.auth
+import requests.exceptions as r_exc
+import six
+
+from delfin import exception
+from delfin.i18n import _
+
+LOG = logging.getLogger(__name__)
+SLOPROVISIONING = 'sloprovisioning'
+U4V_VERSION = '92'
+UCODE_5978 = '5978'
+# HTTP constants
+GET = 'GET'
+POST = 'POST'
+PUT = 'PUT'
+DELETE = 'DELETE'
+STATUS_200 = 200
+STATUS_201 = 201
+STATUS_202 = 202
+STATUS_204 = 204
+
+
+class VMaxRest(object):
+    """Rest class based on Unisphere for VMax Rest API."""
+
+    def __init__(self):
+        self.session = None
+        self.base_uri = None
+        self.user = None
+        self.passwd = None
+        self.verify = None
+        self.cert = None
+        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+
+    def set_rest_credentials(self, array_info, extra_attrib):
+        """Given the array record set the rest server credentials.
+        :param array_info: record
+        :param extra_attrib: record
+        """
+        ip = array_info['host']
+        port = array_info['port']
+        self.user = array_info['username']
+        self.passwd = array_info['password']
+        self.verify = False
+        if extra_attrib:
+            self.verify = extra_attrib.get('verify', False)
+        ip_port = "%(ip)s:%(port)s" % {'ip': ip, 'port': port}
+        self.base_uri = ("https://%(ip_port)s/univmax/restapi" % {
+            'ip_port': ip_port})
+        self.session = self._establish_rest_session()
+
+    def _establish_rest_session(self):
+        """Establish the rest session.
+        :returns: requests.session() -- session, the rest session
+        """
+        LOG.info("Establishing REST session with %(base_uri)s",
+                 {'base_uri': self.base_uri})
+        if self.session:
+            self.session.close()
+        session = requests.session()
```

Comment:
Do we want to keep the session open? Any exception need to close?

Reply:
We are keeping session alive till service restart.